### ISSUE TITLE
EXT_materials_consistency extension proposal

### ISSUE THREAD

**Issue by UX3D-nopper** [KhronosGroup/glTF#1245]
Just talked to the glTF specification board and I want to start the discussion for a material consistency extension.

What is it about?
If you take a glTF model and render it in engine A and render in engine B - having e.g. same environment light and tonemapping algorithm, the render output should be - if this extension is supported and enabled - 99,9% the same.

From a first discussion, probably in most game engines is extension is not required, but e.g. for industry applications, this demand for pixel color perfection is a high demand.

The extension will mainly conatin the math and algortihms, how things should be implemented.

Want to hear your opinions.

**Comment by javagl** [comment]
Just a pointer: Upon request, references had been added to the appendix of the WebGL-PRB reference implementation, in the appendix: https://github.com/KhronosGroup/glTF-WebGL-PBR#appendix They *should* cover much of the computations (disregarding further extensions like HDR, and all to be taken with a grain of salt as long as Environments+Lights are not finalized and considered there accordingly).

Independent of that: I think that it would be helpful to flesh out *where the differences in the appearance come from*. Is it *only* lighting? Is it some differences in the handing of gamma correction, or sRGB/ColorSpace conversions? Maybe differences from World/View space computations? Normalization or Normal Interpolation issues? Are there possibly points in the spec that are still too vague and should be clarified?

<sub>Otherwise, I'd be curious to see who has the balls to say that his implementation is *The Correct One©"*  ;-)</sub>

**Comment by snagy** [comment]
As the person who wrote the appendix, I will emphatically **not** claim that it is "The Correct One", just a suggested starting point for a good implementation. :)

Do we think that the implementations that want repeatable consistency are also properly served by the existing glTF material model?  I know a lot of industry applications which care strongly about stuff like perfect surface color also want deeper transparency/refraction/scattering/anisotropy/etc support than we have in our simple MR/SG model.  Will there be customers for this extension or would it be stuck 'in between' content that doesn't care about consistency (most existing glTF stuff) and content that wants a richer model than we have?

**Comment by donmccurdy** [comment]
> I think that it would be helpful to flesh out where the differences in the appearance come from. Is it only lighting? Is it some differences in the handling of gamma correction, or sRGB/ColorSpace conversions? ...

Strong +1 on this, it would be very helpful to understand where deviations exist, and adapt our approach accordingly. And, as @snagy mentions, to compare that against the features necessary for customers who care about pixel-perfect consistency.

In three.js, there are [known issues ](https://github.com/mrdoob/three.js/issues/13107) with the metal/rough materials that we're planning to improve, to align with other implementations, in any case.

I'm unsure how an extension can/should be employed as a correctness badge... Does the extension contain new information, or just tighten allowed interpretations of the existing spec? If a user tries to load a model with the `EXT_materials_consistency` extension in three.js, and we don't have our metal/rough issues sorted out, must we throw an error rather than render anything? How do we determine which engines really support `EXT_materials_consistency`, and could that same process be applied to the existing spec?

tl;dr, I'm wondering if we can treat this as a conformance issue rather than an extension, and start by identifying the types of problems currently present.

**Comment by donmccurdy** [comment]
OK with me. We have other consistency work ongoing too, but it doesn't currently seem to require an extension to accomplish.

**Comment by UX3D-nopper** [comment]
Agree, think we have to redo everything anyway about consistency.
